altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (C) 2017  AltOS-Rust Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//! Linked list code for the the memory allocator
+//! This is intended for use by the free_list_allocator
+//!
+
+use core::{mem, ptr};
+
+use alignment;
+
+// BlockHeader keeps track of a free block of memory
+#[cfg(not(test))]
+#[repr(C)]
+pub struct BlockHeader {
+    block_size: usize,
+    next_block: *mut BlockHeader,
+}
+
+impl BlockHeader {
+    const fn new(size: usize) -> Self {
+        BlockHeader {
+            block_size: size,
+            next_block: ptr::null_mut(),
+        }
+    }
+}
+
+// FreeList is a linked list which keeps track of free blocks of memory
+// Free blocks are embedded in the free memory itself
+#[cfg(not(test))]
+pub struct FreeList {
+    block_hdr_size: usize,
+    head: *mut BlockHeader,
+}
+
+// These are (trivially) implemented so FreeList objects can be passed between threads.
+unsafe impl Send for FreeList {}
+unsafe impl Sync for FreeList {}
+
+impl FreeList {
+    pub const fn new() -> Self {
+        FreeList {
+            block_hdr_size: 0,
```

Comment:
That makes sense, I'll change this. I think I had it in my mind that size_of would result in a function call, so i put this in to avoid that.